Plugins/Public/kill_tracker/Main.cpp

```diff
@@ -0,0 +1,419 @@
+	}
+}
+
+void __stdcall SendDeathMessage(const wstring& message, uint system, uint clientVictim, uint clientKiller)
+{
+	returncode = DEFAULT_RETURNCODE;
+	if (!clientVictim)
+	{
+		return;
+	}
+
+	returncode = NOFUNCTIONCALL;
+
+	map<float, uint> damageToInflictorMap; // damage is the key instead of value because keys are sorted, used to render top contributors in order
+	set<CPlayerGroup*> killerGroups;
+
+	float totalDamageTaken = 0.0f;
+	PlayerData* pd = nullptr;
+	while (pd = Players.traverse_active(pd))
+	{
+		float damageDone = damageArray[pd->iOnlineID][clientVictim];
+		if (damageDone == 0){
+			continue;
+		}
+		damageToInflictorMap[damageDone] = pd->iOnlineID;
+		killerGroups.insert(pd->PlayerGroup);
+		totalDamageTaken += damageDone;
+	}
+	if (totalDamageTaken == 0.0f)
+	{
+		clearDamageTaken(clientVictim);
+		ProcessNonPvPDeath(message, system);
+		return;
+	}
+
+	wstring victimName = (const wchar_t*)Players.GetActiveCharacterName(clientVictim);
+
+	wstring deathMessage = RandomizeDeathTemplate(clientVictim);
+	deathMessage = ReplaceStr(deathMessage, L"%victim", victimName);
+	wstring assistMessage = L"";
+
+	uint killerCounter = 0;
+
+	for (auto& i = damageToInflictorMap.rbegin(); i != damageToInflictorMap.rend(); i++) // top values at the end
+	{
+		if (i == damageToInflictorMap.rend() || killerCounter >= numberOfKillers)
+		{
+			break;
+		}
+		uint contributionPercentage = static_cast<uint>(round((i->first / totalDamageTaken) * 100));
+		if (contributionPercentage < minimumAssistPercentage)
+		{
+			break;
+		}
+
+		wstring inflictorName = (const wchar_t*)Players.GetActiveCharacterName(i->second);
+		if (killerCounter == 0)
+		{
+			deathMessage = ReplaceStr(deathMessage, L"%killer", inflictorName);
+			deathMessage += L" (" + stows(itos(contributionPercentage)) + L"%)";
+		}
+		else if (killerCounter == 1)
+		{
+			assistMessage = L"Assisted by: " + inflictorName + L" (" + stows(itos(contributionPercentage)) + L"%)";
+		}
+		else
+		{
+			assistMessage += L", " + inflictorName + L" (" + stows(itos(contributionPercentage)) + L"%)";
+		}
+		killerCounter++;
+	}
+
+	deathMessage = L"<TRA data=\"" + killMsgStyle +
+		L"\" mask=\"-1\"/><TEXT>" + XMLText(deathMessage) + L"</TEXT>";
+	if (!assistMessage.empty())
+	{
+		assistMessage = L"<TRA data=\"" + killMsgStyle +
+			L"\" mask=\"-1\"/><TEXT>" + XMLText(assistMessage) + L"</TEXT>";
+	}
+
+	CPlayerGroup* victimGroup = victimGroup = Players[clientVictim].PlayerGroup;
+
+	uint systemId;
+	pub::Player::GetSystem(clientVictim, systemId);
+	Vector victimPos;
+	Matrix victimOri;
+	pub::SpaceObj::GetLocation(Players[clientVictim].iShipID, victimPos, victimOri);
+
+	pd = nullptr;
+	while (pd = Players.traverse_active(pd))
+	{
+		uint playerId = pd->iOnlineID;
+		if (damageArray[playerId][clientVictim] > 0)
+		{
+			HkFMsg(playerId, deathMessage);
+			if(!assistMessage.empty())
+				HkFMsg(playerId, assistMessage);
+			continue;
+		}
+		if (
```

Comment:
When you have an if that looks like _this_ you should document it with a comment